File: leetcode_go/3.Longest_Substring_Without_Repeating_Characters.go
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

func lengthOfLongestSubstring(s string) int {
	i, res := 0, 0
	m := map[string]int{}

	for j := 0; j < len(s); j++ {
		if index, ok := m[string(s[j])]; ok {
			i = max(i, index+1)
		}
		m[string(s[j])] = j
		res = max(res, j-i+1)
	}
	return res
}

func max(a int, b int) int {
	if a >= b {
		return a
	}
	return b
}

File: leetcode_cpp/3.Longest_Substring_Without_Repeating_Characters.cpp
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

class Solution {
 public:
  int lengthOfLongestSubstring(string s) {
    unordered_map<int, int> map;

    int i = 0;
    int res = 0;
    for (int j = 0; j < s.size(); j++) {
      if (map.find(s[j]) != map.end()) {
        i = max(i, map[s[j]] + 1);
      }
      map[s[j]] = j;
      res = max(res, j - i + 1);
    }
    return res;
  }
};